src/devices/BuildHat/Brick.cs

```diff
@@ -0,0 +1,1580 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Device.Gpio;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.IO.Ports;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using BuildHat.Models;
+using Iot.Device.BuildHat.Models;
+using Iot.Device.BuildHat.Motors;
+using Iot.Device.BuildHat.Sensors;
+using SixLabors.ImageSharp;
+using UnitsNet;
+
+namespace Iot.Device.BuildHat
+{
+    /// <summary>
+    /// The main Brick class allowing low level access to motors and sensors
+    /// </summary>
+    public class Brick : IDisposable
+    {
+        private const string FirmwareVersion = "Firmware version: ";
+        private const string FirmwareSignature = "Firmware signature: ";
+        private const string GetFirmwareVersion = "version\r";
+        private const string ButtonReleased = ": button released";
+        private const string ButtonPressed = ": button pressed";
+        private const string PowerFaultError = "power fault";
+
+        private SerialPort _port;
```

Comment:
I was hesitating. If I operate on streams, I'll add the possibility to add as well a constructor to pass a stream.

Reply:
now, I know why I haven't been using streams: ReadlLine()... Almost everything is based of it. And all is text (except firmware update). Using streams would add complexity in this case as I'll have to manually implement all this. I may have missed something here again but seems like the BaseStream is just a Stream, so not text primitives at all.

Reply:
Yes, you would (and should) add a constructor to operate on streams, that's the reason for my suggestion. 

You can wrap a StreamReader/StreamWriter around the Stream, to get the ReadLine()/WriteLine() functions back. I agree, writing that yourself would be ugly.

Reply:
I've been trying with the streams and it's not really working. This may be due to the drivers of my usb to ttl but it's really hugly. I get time outs everywhere, the streams do not properly give the number of available bytes or even if there is the EndOfStream property neither.
This hat can only be used with a serial port, so what is the real core interest to have it with streams? Especially as it does drastically complicate this implementation and requires then quite some try/catch.

Reply:
That's a bit weird, indeed. 

The reason why I prefer Stream is that it is more flexible. It can be forwarded trough another link, written to disk (and replayed) or mocked for unit testing. But if you manage to write unit tests against the current implementation as well, this won't be a blocker for me.

Reply:
For the unit test, I added an internal constructor and used reflection to call internal methods. There is not much to test fully anyway. Mainly one function which makes the math on how to get from one potision to another